File: ZezeJava/ZezeJava/src/main/java/Zeze/Services/GlobalCacheManagerPerf.java
package Zeze.Services;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentSkipListMap;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.LongAdder;
import java.util.concurrent.locks.ReentrantLock;
import Zeze.Net.Protocol;
import Zeze.Util.Task;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

public class GlobalCacheManagerPerf extends ReentrantLock {
	private static final int ACQUIRE_STATE_COUNT = 3;
	private static final Logger logger = LogManager.getLogger(GlobalCacheManagerPerf.class);

	private final String perfName;
	private final AtomicLong serialIdGenerator;
	private long lastSerialId;

	private final ConcurrentHashMap<Protocol<?>, Long> acquires = new ConcurrentHashMap<>();
	private final ConcurrentHashMap<Protocol<?>, Long> reduces = new ConcurrentHashMap<>();

	private final LongAdder[] totalAcquireCounts = new LongAdder[ACQUIRE_STATE_COUNT];
	private final LongAdder[] totalAcquireTimes = new LongAdder[ACQUIRE_STATE_COUNT];
	private final AtomicLong[] maxAcquireTimes = new AtomicLong[ACQUIRE_STATE_COUNT];
	private final LongAdder totalReduceCount = new LongAdder();
	private final LongAdder totalReduceTime = new LongAdder();
	private final AtomicLong maxReduceTime = new AtomicLong();
	@SuppressWarnings("unchecked")
	private final ConcurrentSkipListMap<Long, LongAdder>[] totalAcquireResults = new ConcurrentSkipListMap[ACQUIRE_STATE_COUNT];
	private final ConcurrentSkipListMap<Long, LongAdder> totalReduceResults = new ConcurrentSkipListMap<>();
	private final ConcurrentSkipListMap<String, LongAdder> others = new ConcurrentSkipListMap<>();

	GlobalCacheManagerPerf(@NotNull String perfName, @NotNull AtomicLong serialIdGenerator) {
		this.perfName = perfName;
		this.serialIdGenerator = serialIdGenerator;
		lastSerialId = serialIdGenerator.get();
		for (int i = 0; i < ACQUIRE_STATE_COUNT; i++) {
			totalAcquireCounts[i] = new LongAdder();
			totalAcquireTimes[i] = new LongAdder();
			maxAcquireTimes[i] = new AtomicLong();
			totalAcquireResults[i] = new ConcurrentSkipListMap<>();
		}
		Task.schedule(1000, 1000, this::report);
	}

	void onAcquireBegin(@NotNull Protocol<?> rpc, int state) {
		if (Integer.compareUnsigned(state, ACQUIRE_STATE_COUNT) < 0
				&& acquires.put(rpc, System.nanoTime()) != null)
			logger.warn("onAcquireBegin again");
	}

	void onAcquireEnd(@NotNull Protocol<?> rpc, int state) {
		var beginTime = acquires.remove(rpc);
		if (beginTime != null) {
			var time = System.nanoTime() - beginTime;
			totalAcquireCounts[state].increment();
			totalAcquireTimes[state].add(time);
			var maxAcquireTime = maxAcquireTimes[state];
			long maxTime;
			do
				maxTime = maxAcquireTime.get();
			while (time > maxTime && !maxAcquireTime.compareAndSet(maxTime, time));
			if (rpc.getResultCode() != 0)
				totalAcquireResults[state].computeIfAbsent(rpc.getResultCode(), __ -> new LongAdder()).increment();
		}
	}

	void onReduceBegin(@NotNull Protocol<?> rpc) {
		if (reduces.put(rpc, System.nanoTime()) != null)
			logger.warn("already onReduceBegin: {}", rpc);
	}

	void onReduceCancel(@NotNull Protocol<?> rpc) {
		if (reduces.remove(rpc) == null)
			logger.warn("already onReduceCancel: {}", rpc);
	}

	void onReduceEnd(@NotNull Protocol<?> rpc) {
		var beginTime = reduces.remove(rpc);
		if (beginTime != null) {
			var time = System.nanoTime() - beginTime;
			totalReduceCount.increment();
			totalReduceTime.add(time);
			long maxTime;
			do
				maxTime = maxReduceTime.get();
			while (time > maxTime && !maxReduceTime.compareAndSet(maxTime, time));
			if (rpc.getResultCode() != 0)
				totalReduceResults.computeIfAbsent(rpc.getResultCode(), __ -> new LongAdder()).increment();
		} else
			logger.warn("already onReduceEnd: {}", rpc);
	}

	void onOthers(@NotNull String info) {
		others.computeIfAbsent(info, __ -> new LongAdder()).increment();
	}

	private void report() {
		lock();
		try {
			long curSerialId = serialIdGenerator.get();
			long serialIds = curSerialId - lastSerialId;
			lastSerialId = curSerialId;
			long totalReduceCountSum = totalReduceCount.sumThenReset();
			long acquiresSize = acquires.size();
			long reducesSize = reduces.size();
			long[] totalAcquireCounts0 = new long[ACQUIRE_STATE_COUNT];
			for (int i = 0; i < ACQUIRE_STATE_COUNT; i++)
				totalAcquireCounts0[i] = totalAcquireCounts[i].sumThenReset();

			if ((serialIds | totalReduceCountSum | acquiresSize | reducesSize) == 0) {
				int i = 0;
				for (; i < ACQUIRE_STATE_COUNT; i++) {
					if (totalAcquireCounts0[i] != 0)
						break;
				}
				if (i == ACQUIRE_STATE_COUNT)
					return;
			}

			var sb = new StringBuilder().append("SerialIds = ").append(serialIds).append('\n');
			for (int i = 0; i < ACQUIRE_STATE_COUNT; i++) {
				long count = totalAcquireCounts0[i];
				sb.append("Acquires[").append(i).append("] = ").append(count);
				if (count > 0) {
					sb.append(", ").append(totalAcquireTimes[i].sumThenReset() / count / 1_000).append(" us/acquire, max: ")
							.append(maxAcquireTimes[i].getAndSet(0) / 1_000_000).append(" ms");
				}
				for (var e : totalAcquireResults[i].entrySet())
					sb.append(", r=").append(e.getKey()).append(':').append(e.getValue().sum());
				totalAcquireResults[i].clear();
				sb.append('\n');
			}
			sb.append("Reduces = ").append(totalReduceCountSum);
			if (totalReduceCountSum > 0) {
				sb.append(", ").append(totalReduceTime.sumThenReset() / totalReduceCountSum / 1_000)
						.append(" us/reduce, max: ").append(maxReduceTime.getAndSet(0) / 1_000_000).append(" ms");
				for (var e : totalReduceResults.entrySet())
					sb.append(", r=").append(e.getKey()).append(':').append(e.getValue().sum());
				totalReduceResults.clear();
			}
			sb.append("\nAcquire/Reduce Pendings = ").append(acquiresSize).append(" / ").append(reducesSize).append('\n');
			for (var e : others.entrySet())
				sb.append(e.getKey()).append(" = ").append(e.getValue().sum()).append('\n');
			others.clear();
			var es = Task.getThreadPool();
			if (es instanceof ThreadPoolExecutor) {
				var queueSize = ((ThreadPoolExecutor)es).getQueue().size();
				if (queueSize != 0)
					sb.append("ThreadPoolQueueSize = ").append(queueSize).append('\n');
			}
			logger.info("{}\n{}", perfName, sb.toString());
		} finally {
			unlock();
		}
	}
}

File: Zeze/Services/GlobalCacheManagerPerf.cs

using System.Collections.Concurrent;
using System.Text;
using Zeze.Util;

namespace Zeze.Services.GlobalCacheManager
{
	public class GlobalCacheManagerPerf
	{
		private const int ACQUIRE_STATE_COUNT = 3;
		private static readonly NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();

		private readonly Util.AtomicLong serialIdGenerator;
		private long lastSerialId;

		private readonly ConcurrentDictionary<Acquire, long> acquires = new();
		private readonly ConcurrentDictionary<Reduce, long> reduces = new();

		private readonly AtomicLong[] totalAcquireCounts = new AtomicLong[ACQUIRE_STATE_COUNT];
		private readonly AtomicLong[] totalAcquireTimes = new AtomicLong[ACQUIRE_STATE_COUNT];
		private readonly AtomicLong[] maxAcquireTimes = new AtomicLong[ACQUIRE_STATE_COUNT];
		private readonly AtomicLong totalReduceCount = new();
		private readonly AtomicLong totalReduceTime = new();
		private readonly AtomicLong maxReduceTime = new();

		public GlobalCacheManagerPerf(Util.AtomicLong serialIdGenerator)
		{
			this.serialIdGenerator = serialIdGenerator;
			lastSerialId = serialIdGenerator.Get();
			for (int i = 0; i < ACQUIRE_STATE_COUNT; i++)
			{
				totalAcquireCounts[i] = new AtomicLong();
				totalAcquireTimes[i] = new AtomicLong();
				maxAcquireTimes[i] = new AtomicLong();
			}
			Scheduler.Schedule(Report, 1000, 1000);
		}

		public void OnAcquireBegin(Acquire rpc)
		{
			if ((rpc.Argument.State & 0xffff_ffffL) < ACQUIRE_STATE_COUNT)
				acquires[rpc] = Time.NanoTime();
		}

		public void OnAcquireEnd(Acquire rpc)
		{
			if (acquires.TryRemove(rpc, out var beginTime))
			{
				var time = Time.NanoTime() - beginTime;
				totalAcquireCounts[rpc.Argument.State].IncrementAndGet();
				totalAcquireTimes[rpc.Argument.State].AddAndGet(time);
				var maxAcquireTime = maxAcquireTimes[rpc.Argument.State];
				long maxTime;
				do
					maxTime = maxAcquireTime.Get();
				while (time > maxTime && !maxAcquireTime.CompareAndSet(maxTime, time));
			}
		}

		public void OnReduceBegin(Reduce rpc)
		{
			reduces[rpc] = Time.NanoTime();
		}

		public void OnReduceCancel(Reduce rpc)
		{
			reduces.TryRemove(rpc, out _);
		}

		public void OnReduceEnd(Reduce rpc)
		{
			if (reduces.TryRemove(rpc, out var beginTime))
			{
				var time = Time.NanoTime() - beginTime;
				totalReduceCount.IncrementAndGet();
				totalReduceTime.AddAndGet(time);
				long maxTime;
				do
					maxTime = maxReduceTime.Get();
				while (time > maxTime && !maxReduceTime.CompareAndSet(maxTime, time));
			}
		}

		private void Report(SchedulerTask ThisTask)
		{
			long curSerialId = serialIdGenerator.Get();
			long serialIds = curSerialId - lastSerialId;
			lastSerialId = curSerialId;

			if ((serialIds | totalReduceCount.Get() | (long)acquires.Count | (long)reduces.Count) == 0)
			{
				int i = 0;
				for (; i < ACQUIRE_STATE_COUNT; i++)
				{
					if (totalAcquireCounts[i].Get() != 0)
						break;
				}
				if (i == ACQUIRE_STATE_COUNT)
					return;
			}

			var sb = new StringBuilder().Append("SerialIds = ").Append(serialIds).Append('\n');
			for (int i = 0; i < ACQUIRE_STATE_COUNT; i++)
			{
				long c = totalAcquireCounts[i].Get();
				sb.Append("Acquires[").Append(i).Append("] = ").Append(c);
				if (c > 0)
				{
					sb.Append(", ").Append(totalAcquireTimes[i].Get() / c / 1_000).Append(" us/acquire, max: ")
							.Append(maxAcquireTimes[i].Get() / 1_000_000).Append(" ms");
				}
				sb.Append('\n');
			}
			long count = totalReduceCount.Get();
			sb.Append("Reduces = ").Append(count);
			if (count > 0)
			{
				sb.Append(", ").Append(totalReduceTime.Get() / count / 1_000).Append(" us/reduce, max: ")
						.Append(maxReduceTime.Get() / 1_000_000).Append(" ms");
			}
			sb.Append("\nAcquire/Reduce Pendings = ").Append(acquires.Count).Append(" / ")
					.Append(reduces.Count).Append('\n');
			logger.Info($"\n{sb}");

			for (int i = 0; i < ACQUIRE_STATE_COUNT; i++)
			{
				totalAcquireCounts[i].GetAndSet(0);
				totalAcquireTimes[i].GetAndSet(0);
				maxAcquireTimes[i].GetAndSet(0);
			}
			totalReduceCount.GetAndSet(0);
			totalReduceTime.GetAndSet(0);
			maxReduceTime.GetAndSet(0);
		}
	}
}
